core/src/main/java/org/infinispan/util/concurrent/CompletionStages.java

```diff
@@ -0,0 +1,201 @@
+    * Returns a CompletionStage that also can be composed of many other CompletionStages. A stage can compose another
+    * stage in it by invoking the {@link AggregateCompletionStage#dependsOn(CompletionStage)} method passing in the
+    * CompletionStage. After all stages this composition stage depend upon have been added, the
+    * {@link AggregateCompletionStage#freeze()} should be invoked so that the AggregateCompletionStage can finally
+    * complete when all of the stages it depends upon complete.
+    * <p>
+    * If any stage this depends upon fails the returned stage will contain the Throwable from one of the stages.
+    * @return composed completion stage
+    */
+   public static AggregateCompletionStage<Void> aggregateCompletionStage() {
+      return new VoidAggregateCompletionStage();
+   }
+
+   /**
+    * Same as {@link #aggregateCompletionStage()} except that when this stage completes normally it will return
+    * the value provided.
+    * @param valueToReturn value to return to future stage compositions
+    * @param <R> the type of the value
+    * @return composed completion stage that returns the value upon normal completion
+    */
+   public static <R> AggregateCompletionStage<R> composedCompletionStage(R valueToReturn) {
+      return new ValueAggregateCompletionStage<>(valueToReturn);
+   }
+
+   /**
+    * Returns if the provided {@link CompletionStage} has already completed normally, that is not due to an exception.
+    * @param stage stage to check
+    * @return if the stage is completed normally
+    */
+   public static boolean isCompleteSuccessfully(CompletionStage<?> stage) {
+      CompletableFuture<?> future = stage.toCompletableFuture();
+      return future.isDone() && !future.isCompletedExceptionally();
+   }
+
+   /**
+    * Returns the result value when complete, or throws an (unchecked) exception if completed exceptionally.
+    * To better conform with the use of common functional forms, if a computation involved in the completion of this
+    * CompletionStage threw an exception, this method throws an (unchecked) CompletionException with the underlying
+    * exception as its cause.
+    * @param stage stage to wait on
+    * @param <R> the type in the stage
+    * @return the result value
+    * @throws CompletionException if this stage completed exceptionally or a completion computation threw an exception
+    */
+   public static <R> R join(CompletionStage<R> stage) {
+      return stage.toCompletableFuture().join();
+   }
+
+   /**
+    * Shorthand for <code>CompletionStages.join(CompletionStages.allOf(stage1, stage2))</code> but doesn't return a value
+    * @param stage1
+    * @param stage2
+    * @throws CompletionException if either future completed exceptionally or a completion computation threw an exception
+    */
+   public static void await(CompletionStage<?> stage1, CompletionStage<?> stage2) {
+      join(allOf((CompletionStage<Void>) stage1, (CompletionStage<Void>) stage2));
```

Comment:
Does `allOf` really need the parameters to be `CompletionStage<Void>`?

Reply:
The problem is that if it isn't Void then I can't just return it directly. So sadly this one does require Void for both.